File: src/TaxAuthorityProvider/TaxAuthorities/Slovenia/ReportGeneration/KDVP/XML_Doh_KDVP.py
from typing import Sequence

from lxml import etree

import src.ConfigurationProvider.Configuration as c
import src.Core.Schemas.ProcessedGenericFormats as pgf
import src.TaxAuthorityProvider.Schemas.Configuration as tc
import src.TaxAuthorityProvider.TaxAuthorities.Slovenia.ReportGeneration.KDVP.Common as common
import src.TaxAuthorityProvider.TaxAuthorities.Slovenia.Schemas.ReportTypes as rt
import src.TaxAuthorityProvider.TaxAuthorities.Slovenia.Schemas.Schemas as ss


def generateXmlReport(
    reportConfig: tc.TaxAuthorityConfiguration,
    userConfig: c.TaxPayerInfo,
    documentType: rt.EDavkiDocumentWorkflowType,
    data: Sequence[pgf.UnderlyingGrouping],
    templateEnvelope: etree._Element,
) -> etree._Element:
    convertedTrades = common.convertTradesToKdvpItems(reportConfig, data)

    nsmap = templateEnvelope.nsmap
    nsmap[None] = "http://edavki.durs.si/Documents/Schemas/Doh_KDVP_9.xsd"
    envelope = etree.Element(templateEnvelope.tag, attrib=templateEnvelope.attrib, nsmap=nsmap)
    envelope[:] = templateEnvelope[:]

    body = etree.SubElement(envelope, "body")
    etree.SubElement(body, etree.QName(nsmap["edp"], "bodyContent"))

    Doh_KDVP = etree.SubElement(body, "Doh_KDVP")
    KDVP = etree.SubElement(Doh_KDVP, "KDVP")
    etree.SubElement(KDVP, "DocumentWorkflowID").text = documentType.value
    etree.SubElement(KDVP, "Year").text = reportConfig.fromDate.format("YYYY")
    etree.SubElement(KDVP, "PeriodStart").text = reportConfig.fromDate.format("YYYY-MM-DD")
    etree.SubElement(KDVP, "PeriodEnd").text = reportConfig.toDate.shift(days=-1).format("YYYY-MM-DD")
    etree.SubElement(KDVP, "IsResident").text = str(userConfig.countryID == "SI").lower()
    etree.SubElement(KDVP, "SecurityCount").text = "0"
    etree.SubElement(KDVP, "SecurityShortCount").text = "0"
    etree.SubElement(KDVP, "SecurityWithContractCount").text = "0"
    etree.SubElement(KDVP, "SecurityWithContractShortCount").text = "0"
    etree.SubElement(KDVP, "ShareCount").text = "0"
    # etree.SubElement(KDVP, "CountryOfResidenceID").text = config.countryID
    # etree.SubElement(KDVP, "TelephoneNumber").text = "telefonska"
    # etree.SubElement(KDVP, "Email").text = "email"

    for KDVP_item_entry in convertedTrades:
        KDVP_ITEM = etree.SubElement(Doh_KDVP, "KDVPItem")

        etree.SubElement(KDVP_ITEM, "InventoryListType").text = KDVP_item_entry.InventoryListType.value
        etree.SubElement(KDVP_ITEM, "HasForeignTax").text = str(KDVP_item_entry.HasForeignTax).lower()
        if KDVP_item_entry.ForeignTax is not None:
            etree.SubElement(KDVP_ITEM, "ForeignTax").text = str(KDVP_item_entry.ForeignTax.__round__(5))

        if KDVP_item_entry.FTCountryID is not None:
            etree.SubElement(KDVP_ITEM, "FTCountryID").text = KDVP_item_entry.FTCountryID
        if KDVP_item_entry.FTCountryName is not None:
            etree.SubElement(KDVP_ITEM, "FTCountryName").text = KDVP_item_entry.FTCountryName
        etree.SubElement(KDVP_ITEM, "HasLossTransfer").text = str(KDVP_item_entry.HasLossTransfer or False).lower()
        etree.SubElement(KDVP_ITEM, "ForeignTransfer").text = str(KDVP_item_entry.ForeignTransfer or False).lower()
        etree.SubElement(KDVP_ITEM, "TaxDecreaseConformance").text = str(KDVP_item_entry.TaxDecreaseConformance or False).lower()

        if KDVP_item_entry.InventoryListType == ss.EDavkiTradeSecurityType.SECURITY:
            for securityEntry in KDVP_item_entry.Items:
                entry = etree.SubElement(KDVP_ITEM, "Securities")

                etree.SubElement(entry, "ISIN").text = securityEntry.ISIN
                etree.SubElement(entry, "Code").text = securityEntry.Code
                # etree.SubElement(entry, "Name").text = securityEntry.Name
                etree.SubElement(entry, "IsFond").text = str(securityEntry.IsFund).lower()

                if securityEntry.Resolution is not None:
                    etree.SubElement(entry, "Resolution").text = securityEntry.Resolution

                if securityEntry.ResolutionDate is not None:
                    etree.SubElement(entry, "ResolutionDate").text = securityEntry.ResolutionDate.format("YYYY-MM-DD")

                for id, entryLine in enumerate(securityEntry.Events):
                    row = etree.SubElement(entry, "Row")

                    etree.SubElement(row, "ID").text = str(id)

                    if isinstance(
                        entryLine,
                        ss.EDavkiTradeReportSecurityLineGenericEventBought,
                    ):
                        purchase = etree.SubElement(row, "Purchase")
                        etree.SubElement(purchase, "F1").text = entryLine.BoughtOn.format("YYYY-MM-DD")
                        etree.SubElement(purchase, "F2").text = entryLine.GainType.value
                        etree.SubElement(purchase, "F3").text = str(entryLine.Quantity.__round__(5))
                        etree.SubElement(purchase, "F4").text = str(entryLine.PricePerUnit.__round__(5))
                        etree.SubElement(purchase, "F5").text = str(
                            entryLine.InheritanceAndGiftTaxPaid.__round__(5) if entryLine.InheritanceAndGiftTaxPaid is not None else "0"
                        )
                        etree.SubElement(purchase, "F11").text = str(
                            entryLine.BaseTaxReduction.__round__(5) if entryLine.BaseTaxReduction is not None else "0"
                        )

                    if isinstance(entryLine, ss.EDavkiTradeReportSecurityLineGenericEventSold):
                        sale = etree.SubElement(row, "Sale")
                        etree.SubElement(sale, "F6").text = entryLine.SoldOn.format("YYYY-MM-DD")
                        etree.SubElement(sale, "F7").text = str(entryLine.Quantity.__round__(5))
                        etree.SubElement(sale, "F9").text = str(entryLine.PricePerUnit.__abs__().__round__(5))
                        etree.SubElement(sale, "F10").text = str(entryLine.SatisfiesTaxBasisReduction).lower()

    return envelope

File: lib/TaxAuthorities/Slovenia/ReportGeneration/Kdvp/XmlDohKdvp.ts
import { XMLBuilder } from "fast-xml-parser";
import type { TaxAuthorityConfiguration } from "@brrr/TaxAuthorities/ConfigurationProvider.ts";
import type { TaxPayerInfo } from "@brrr/TaxAuthorities/ConfigurationProvider.ts";
import type { EDavkiDocumentWorkflowType } from "@brrr/TaxAuthorities/Slovenia/Schemas/ReportTypes.ts";
import {
	type EDavkiGenericTradeReportItem,
	type EDavkiTradeReportSecurityLineGenericEventBought,
	type EDavkiTradeReportSecurityLineGenericEventSold,
	EDavkiTradeSecurityType,
} from "@brrr/TaxAuthorities/Slovenia/Schemas/Schemas.ts";

function buildXmlTree(
	reportConfig: TaxAuthorityConfiguration,
	userConfig: TaxPayerInfo,
	documentType: EDavkiDocumentWorkflowType,
	kdvpItems: EDavkiGenericTradeReportItem[],
	envelopeAttribs: Record<string, string>,
): unknown {
	const edp = "http://edavki.durs.si/Documents/Schemas/EDP-Common-1.xsd";

	const kdvpItemNodes = kdvpItems.map((kdvpItem) => {
		const securitiesNodes = kdvpItem.inventoryListType === EDavkiTradeSecurityType.SECURITY
			? kdvpItem.items.map((secEntry) => {
				const rowNodes = secEntry.events.map((entryLine, id) => {
					if (entryLine.kind === "Bought") {
						const bought = entryLine as EDavkiTradeReportSecurityLineGenericEventBought;
						const purchaseFields: Record<string, string> = {
							F1: bought.boughtOn.toFormat("yyyy-MM-dd"),
							F2: bought.gainType,
							F3: String(round(bought.quantity, 5)),
							F4: String(round(bought.pricePerUnit, 5)),
							F5: String(round(bought.inheritanceAndGiftTaxPaid ?? 0, 5)),
						};
						if (bought.baseTaxReduction !== null) {
							purchaseFields["F11"] = String(round(bought.baseTaxReduction, 5));
						}
						return { ID: String(id), Purchase: purchaseFields };
					} else {
						const sold = entryLine as EDavkiTradeReportSecurityLineGenericEventSold;
						return {
							ID: String(id),
							Sale: {
								F6: sold.soldOn.toFormat("yyyy-MM-dd"),
								F7: String(round(-sold.quantity, 5)),
								F9: String(round(Math.abs(sold.pricePerUnit), 5)),
								F10: String(sold.satisfiesTaxBasisReduction).toLowerCase(),
							},
						};
					}
				});

				const securitiesObj: Record<string, unknown> = {
					ISIN: secEntry.isin,
					Code: secEntry.code ?? "",
					IsFond: String(secEntry.isFund).toLowerCase(),
					Row: rowNodes,
				};
				if (secEntry.resolution !== null) {
					securitiesObj["Resolution"] = secEntry.resolution;
				}
				if (secEntry.resolutionDate !== null) {
					securitiesObj["ResolutionDate"] = secEntry.resolutionDate!.toFormat("yyyy-MM-dd");
				}
				return securitiesObj;
			})
			: [];

		const kdvpItemObj: Record<string, unknown> = {
			InventoryListType: kdvpItem.inventoryListType,
			HasForeignTax: String(kdvpItem.hasForeignTax).toLowerCase(),
			HasLossTransfer: String(kdvpItem.hasLossTransfer ?? false).toLowerCase(),
			ForeignTransfer: String(kdvpItem.foreignTransfer ?? false).toLowerCase(),
			TaxDecreaseConformance: String(kdvpItem.taxDecreaseConformance ?? false).toLowerCase(),
		};
		if (kdvpItem.foreignTax !== null) {
			kdvpItemObj["ForeignTax"] = String(round(kdvpItem.foreignTax, 5));
		}
		if (kdvpItem.ftCountryId !== null) {
			kdvpItemObj["FTCountryID"] = kdvpItem.ftCountryId;
		}
		if (kdvpItem.ftCountryName !== null) {
			kdvpItemObj["FTCountryName"] = kdvpItem.ftCountryName;
		}
		if (securitiesNodes.length > 0) {
			kdvpItemObj["Securities"] = securitiesNodes;
		}
		return kdvpItemObj;
	});

	return {
		"?xml": { "@_version": "1.0", "@_encoding": "UTF-8" },
		Envelope: {
			"@_xmlns:edp": edp,
			"@_xmlns:xs": "http://www.w3.org/2001/XMLSchema",
			...envelopeAttribs,
			"edp:Header": {
				"edp:taxpayer": {
					"edp:taxNumber": userConfig.taxNumber,
					"edp:taxpayerType": userConfig.taxpayerType,
					"edp:name": userConfig.name,
					"edp:address1": userConfig.address1,
					"edp:address2": userConfig.address2 ?? "",
					"edp:city": userConfig.city,
					"edp:postNumber": userConfig.postNumber,
					"edp:postName": userConfig.postName,
				},
				"edp:Workflow": {
					"edp:DocumentWorkflowID": documentType,
				},
			},
			"edp:AttachmentList": {},
			"edp:Signatures": {},
			body: {
				"edp:bodyContent": {},
				Doh_KDVP: {
					KDVP: {
						DocumentWorkflowID: documentType,
						Year: reportConfig.fromDate.toFormat("yyyy"),
						PeriodStart: reportConfig.fromDate.toFormat("yyyy-MM-dd"),
						PeriodEnd: reportConfig.toDate.minus({ days: 1 }).toFormat("yyyy-MM-dd"),
						IsResident: String(userConfig.countryId === "SI").toLowerCase(),
						SecurityCount: "0",
						SecurityShortCount: "0",
						SecurityWithContractCount: "0",
						SecurityWithContractShortCount: "0",
						ShareCount: "0",
					},
					KDVPItem: kdvpItemNodes,
				},
			},
		},
	};
}

function round(value: number, decimals: number): number {
	const factor = Math.pow(10, decimals);
	return Math.round(value * factor) / factor;
}

export function generateXmlReport(
	reportConfig: TaxAuthorityConfiguration,
	userConfig: TaxPayerInfo,
	documentType: EDavkiDocumentWorkflowType,
	kdvpItems: EDavkiGenericTradeReportItem[],
): string {
	const builder = new XMLBuilder({
		ignoreAttributes: false,
		attributeNamePrefix: "@_",
		format: true,
		suppressEmptyNode: false,
	});

	const tree = buildXmlTree(reportConfig, userConfig, documentType, kdvpItems, {});
	return builder.build(tree) as string;
}
